consensus/fast-sync/src/fast_sync.rs

```diff
@@ -0,0 +1,217 @@
+use crate::{hash_of_hashes, BlockId, HashOfHashes};
+
+#[allow(unused_imports)]
+use hex_literal::hex;
+
+use std::{
+    cmp,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::Service;
+
+#[cfg(not(test))]
+static HASHES_OF_HASHES: &[HashOfHashes] = &include!("./data/hashes_of_hashes");
+
+#[cfg(not(test))]
+const BATCH_SIZE: usize = 512;
+
+#[cfg(test)]
+static HASHES_OF_HASHES: &[HashOfHashes] = &[
+    hex!("3fdc9032c16d440f6c96be209c36d3d0e1aed61a2531490fe0ca475eb615c40a"),
+    hex!("0102030405060708010203040506070801020304050607080102030405060708"),
+    hex!("0102030405060708010203040506070801020304050607080102030405060708"),
+];
+
+#[cfg(test)]
+const BATCH_SIZE: usize = 4;
+
+#[inline]
+fn max_height() -> u64 {
+    (HASHES_OF_HASHES.len() * BATCH_SIZE) as u64
+}
+
+pub enum FastSyncRequest {
+    ValidateHashes {
+        start_height: u64,
+        block_ids: Vec<BlockId>,
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ValidBlockId(BlockId);
+
+fn valid_block_ids(block_ids: &[BlockId]) -> Vec<ValidBlockId> {
+    block_ids.iter().map(|b| ValidBlockId(*b)).collect()
+}
+
+#[derive(Debug, PartialEq)]
+pub enum FastSyncResponse {
+    ValidateHashes {
+        validated_hashes: Vec<ValidBlockId>,
+        unknown_hashes: Vec<BlockId>,
+    },
+}
+
+#[derive(Debug, PartialEq)]
+pub enum FastSyncError {
+    InvalidStartHeight, // start_height not a multiple of BATCH_SIZE
+    Mismatch,           // hash does not match
+    NothingToDo,        // no complete batch to check
+    OutOfRange,         // start_height too high
+}
+
+#[allow(dead_code)]
+pub struct FastSyncService<C> {
+    context_svc: C,
+}
+
+impl<C> FastSyncService<C>
+where
+    C: Service<FastSyncRequest, Response = FastSyncResponse, Error = FastSyncError>
+        + Clone
+        + Send
+        + 'static,
+{
+    #[allow(dead_code)]
+    pub(crate) fn new(context_svc: C) -> FastSyncService<C> {
+        FastSyncService { context_svc }
+    }
+}
+
+impl<C> Service<FastSyncRequest> for FastSyncService<C>
+where
+    C: Service<FastSyncRequest, Response = FastSyncResponse, Error = FastSyncError>
+        + Clone
+        + Send
+        + 'static,
+    C::Future: Send + 'static,
+{
+    type Response = FastSyncResponse;
+    type Error = FastSyncError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: FastSyncRequest) -> Self::Future {
+        Box::pin(async move {
+            match req {
+                FastSyncRequest::ValidateHashes {
+                    start_height,
+                    block_ids,
+                } => validate_hashes(start_height, &block_ids).await,
+            }
+        })
+    }
```

Comment:
```suggestion
    fn call(&mut self, FastSyncRequest::ValidateHashes { start_height, block_ids }: FastSyncRequest) -> Self::Future {
        Box::pin(async move {
            validate_hashes(start_height, &block_ids).await
        })
    }
```
@Boog900

Reply:
There are going to be other requests + I don't really like destructuring in the function arguments

Reply:
We are about to implement the second enum variant, `FastSyncRequest::ValidateBlock`, in the follow-up pr. So this change would just get undone then.